### PR TITLE
Return a hal_device_info_riscv_t from cpu hal/remote hal

### PR DESCRIPTION
**Pull request by coldav**
# Overview

Return a hal_device_info_riscv_t from cpu hal/remote hal

# Reason for change

Allow in-tree risc-v to support cpu hal/remote HAL

# Description of change

Change cpu hal to return a reference to the device info.
Fix assert when vlen = 0 in mux kernels.
Add link_shared to hal arguments.
Add the examples/hals directory to the list searched

### CODE REVIEW

File: modules/mux/targets/riscv/external/CMakeLists.txt [hunk 1]
```diff
@@ -23,6 +23,8 @@ set(RISCV_TOOLCHAIN_DIR "${CA_LLVM_INSTALL_DIR}" CACHE PATH
 
 # Iterate through all 'hal_*' sub-directories.
 file(GLOB HAL_DIR_NAMES FOLLOW_SYMLINKS LIST_DIRECTORIES true "hal_*")
+file(GLOB HAL_DIR_NAMES_EXAMPLES FOLLOW_SYMLINKS LIST_DIRECTORIES true ${PROJECT_SOURCE_DIR}/examples/hals/hal_*)
```
Comment (hvdijk) [inline modules/mux/targets/riscv/external/CMakeLists.txt:26]: We talked about possibly doing
```suggestion
file(GLOB HAL_DIR_NAMES_EXAMPLES FOLLOW_SYMLINKS LIST_DIRECTORIES true hal_* ${PROJECT_SOURCE_DIR}/examples/hals/hal_*)
```

File: modules/mux/targets/riscv/source/kernel.cpp [hunk 2]
```diff
@@ -50,10 +47,19 @@ cargo::expected<riscv::kernel_s *, mux_result_t> kernel_s::create(
...
     variant.sub_group_size = meta.sub_group_size.getKnownMinValue();
     if (meta.sub_group_size.isScalable()) {
+      // Assert non-zero as sub_group_size is unknowable without vlen != 0
+      assert(
+          hal_device_info->vlen &&
+          "vlen must be known at runtime to calculate scalable subgroup width");
       variant.sub_group_size *= real_vscale;
     }
     variant.min_work_width = meta.min_work_item_factor.getKnownMinValue();
+    variant.min_work_width = meta.min_work_item_factor.getKnownMinValue();
```
Comment (hvdijk) [inline modules/mux/targets/riscv/source/kernel.cpp:57]: Accidental duplication?

File: modules/mux/targets/riscv/source/kernel.cpp [hunk 3]
```diff
@@ -50,10 +47,19 @@ cargo::expected<riscv::kernel_s *, mux_result_t> kernel_s::create(
...
     if (meta.min_work_item_factor.isScalable()) {
+      // Assert non-zero as sub_group_size is unknowable without vlen != 0
+      assert(
+          hal_device_info->vlen &&
+          "vlen must be known at runtime to calculate scalable subgroup width");
```
Comment (hvdijk) [inline modules/mux/targets/riscv/source/kernel.cpp:62]: There is another use of `real_vscale` below that also needs the same assert. (Is there perhaps a way we can write that assert only once? Maybe make `real_scale` a lambda that includes the assert?)

File: modules/mux/targets/riscv/external/CMakeLists.txt (hunk 1 above)
Comment (coldav) [reply modules/mux/targets/riscv/external/CMakeLists.txt]: good idea

File: modules/mux/targets/riscv/source/kernel.cpp (hunk 2 above)
Comment (coldav) [reply modules/mux/targets/riscv/source/kernel.cpp]: ta

File: modules/mux/targets/riscv/source/kernel.cpp (hunk 3 above)
Comment (coldav) [reply modules/mux/targets/riscv/source/kernel.cpp]: i've tried something similar

**Review by hvdijk — APPROVED** [review]

**Pull request merged**